File: ts/step3-method-sep/massiveRace.ts
export class MassiveRace {
    private static readonly MIN_CAR_COUNT = 2;
    private static readonly MAX_NAME_LENGTH = 5;
    private static readonly MOVE_THRESHOLD = 4;
    private static readonly MIN_ROUND = 1;

    public race(namesInput: string, rounds: number, randomValues: number[]): string {
        const names = this.parseAndValidateNames(namesInput);
        this.validateRounds(rounds);

        const positions = this.initializePositions(names.length);
        this.playRounds(names, rounds, randomValues, positions);

        const winners = this.findWinners(names, positions);
        return this.formatResult(winners);
    }


    private parseAndValidateNames(namesInput: string): string[] {
        if (!namesInput || namesInput.length === 0) {
            throw new Error('자동차 이름을 입력해 주세요.');
        }

        const names = namesInput.split(',').map(name => name.trim());

        if (names.length < MassiveRace.MIN_CAR_COUNT) {
            throw new Error('자동차는 2대 이상이어야 합니다.');
        }

        for (const name of names) {
            if (name.length === 0) {
                throw new Error('자동차 이름은 비어있을 수 없습니다.');
            }

            if (name.length > MassiveRace.MAX_NAME_LENGTH) {
                throw new Error('자동차 이름은 5자를 초과할 수 없습니다.');
            }
        }

        return names;
    }

    private validateRounds(rounds: number): void {
        if (rounds < MassiveRace.MIN_ROUND) {
            throw new Error('시도 횟수는 1 이상이어야 합니다.');
        }
    }


    private initializePositions(count: number): number[] {
        return new Array(count).fill(0);
    }

    private playRounds(
        names: string[],
        rounds: number,
        randomValues: number[],
        positions: number[]
    ): void {
        let randomIndex = 0;

        for (let round = 0; round < rounds; round++) {
            for (let i = 0; i < names.length; i++) {
                if (randomValues[randomIndex] >= MassiveRace.MOVE_THRESHOLD) {
                    positions[i]++;
                }
                randomIndex++;
            }
        }
    }

    private findWinners(names: string[], positions: number[]): string[] {
        const maxPosition = Math.max(...positions);
        const winners: string[] = [];

        for (let i = 0; i < names.length; i++) {
            if (positions[i] === maxPosition) {
                winners.push(names[i]);
            }
        }

        return winners;
    }

    private formatResult(winners: string[]): string {
        return winners.join(', ') + '가 최종 우승했습니다.';
    }
}

File: java/step3-method-sep/src/main/java/badcode/MassiveRace.java
package badcode;

import java.util.ArrayList;
import java.util.List;

public class MassiveRace {

    private static final int MIN_CAR_COUNT = 2;
    private static final int MAX_NAME_LENGTH = 5;
    private static final int MOVE_THRESHOLD = 4;

    public String race(String namesInput, int rounds, int[] randomValues) {
        validateInput(namesInput, rounds);
        List<String> names = parseNames(namesInput);
        int[] positions = playRounds(rounds, randomValues, names);
        List<String> winners = findWinner(positions, names);
        return formatResult(winners);
    }

    private static String formatResult(List<String> winners) {
        return String.join(", ", winners) + "가 최종 우승했습니다.";
    }

    private static List<String> findWinner(int[] positions, List<String> names) {
        int maxPosition = 0;
        for (int position : positions) {
            if (position > maxPosition) {
                maxPosition = position;
            }
        }
        List<String> winners = new ArrayList<>();
        for (int i = 0; i < names.size(); i++) {
            if (positions[i] == maxPosition) {
                winners.add(names.get(i));
            }
        }
        return winners;
    }

    private static int[] playRounds(int rounds, int[] randomValues, List<String> names) {
        int[] positions = new int[names.size()];
        int randomIndex = 0;
        for (int round = 0; round < rounds; round++) {
            for (int i = 0; i < names.size(); i++) {
                if (randomValues[randomIndex] >= MOVE_THRESHOLD) {
                    positions[i]++;
                }
                randomIndex++;
            }
        }
        return positions;
    }

    private static List<String> parseNames(String namesInput) {
        String[] nameArray = namesInput.split(",");
        List<String> names = new ArrayList<>();
        for (String name : nameArray) {
            names.add(name.trim());
        }
        return names;
    }

    private static void validateInput(String namesInput, int rounds) {
        if (namesInput == null || namesInput.isEmpty()) {
            throw new IllegalArgumentException("자동차 이름을 입력해 주세요.");
        }
        String[] nameArray = namesInput.split(",");
        if (nameArray.length < MIN_CAR_COUNT) {
            throw new IllegalArgumentException("자동차는 2대 이상이어야 합니다.");
        }
        for (String name : nameArray) {
            String trimmed = name.trim();
            if (trimmed.isEmpty()) {
                throw new IllegalArgumentException("자동차 이름은 비어있을 수 없습니다.");
            }
            if (trimmed.length() > MAX_NAME_LENGTH) {
                throw new IllegalArgumentException("자동차 이름은 5자를 초과할 수 없습니다.");
            }
        }
        if (rounds <= 0) {
            throw new IllegalArgumentException("시도 횟수는 1 이상이어야 합니다.");
        }
    }
}
